Async API, leverage rq+redis, use docker-compose for service setup

awelzel opened the pull request.

awelzel committed:
Stream remote builds and compute sha256 on the fly

awelzel committed:
pre-commit: Add isort

awelzel committed:
Split hmac request checking into function, allow configurable build urls

And switch to use exceptions rather than return value based

awelzel committed:
Add Python based test client

awelzel committed:
Snapshot

* Make /zeek asynchronous using rq and redis
* Add tiny-benchmark.sh for running micro-benchmarks
* Allow specifying tests in config.yml
* Splits API, test execution and storage into various files
* Need to make the original pcap test case work again.

awelzel commented on ci-based/config.yml:
```diff
@@ -23,3 +34,30 @@ CPU_SET:
+
+  - id: micro-table-ops-copy
+    bench_command: /benchmarker/scripts/tiny-benchmark.sh
+    bench_args: -D -b microbenchmarks/table-ops/copy.zeek
```
Hey @timwoj - I suspect this PR will be somewhat overwhelming, maybe the config.yml is a good place to get a rough idea for the configuration of multiple jobs. The other is storage.py where the database/persistence part part has initially moved.

awelzel committed:
Add logging microbenchmarks

awelzel committed:
Add function-call and pattern micro benchmarks

awelzel committed:
Increase rq default timeout

Default is 180 seconds which is too small for running multiple jobs.

awelzel committed:
Split benchmarker.py into zeek_benchmarker.app

Use the create_app() pattern and current_app context. This makes
it easier to test the application via Flask's test_client(). See
the new tests/test_app.py file.

awelzel committed:
Tweak microbenchmark n values

Make them run a bit longer.

awelzel committed:
client: Support /broker, add optional commit parameter

awelzel committed:
Make /broker asynchronous, too.

This is untested and not re-worked due to broker builds not
containing the needed broker-cluster-benchmark file.

app.py is now quite shallow

awelzel committed:
Unpack build.tgz once

Instead of unpacking the build.tgz archive, unpack it once into
a volume (zeek_install_data) and mount the contents of that volume
at /zeek/install.

This reduces runtime by not redundantly extracting build.tgz and
also simplifies the setup a bit as we do not need to set ZEEKPATH
anymore (/zeek/install is where Zeek is configured to be installed).

awelzel committed:
microbenchmarks: Add complex-key-value-2, timing tweaks, WRITER_NONE

awelzel committed:
Fixup API tests

awelzel committed:
tasks: Use docker library for running unpack container

...instead of subprocess.run() of /usr/bin/docker

awelzel committed:
pre-commit-config: Add isort

awelzel committed:
Add GitHub workflow for pre-commit

awelzel committed:
client.py: Move into tools, fix formatting

awelzel committed:
Add GitHub action running tests

awelzel committed:
storage: UNIXEPOCH is too new

Not available on Python 3.10 or Python 3.11 in GitHub actions :-(

timwoj commented on ci-based/settings.py:
```diff
@@ -0,0 +1,32 @@
+# If you want to use a dictConfig
+# <https://docs.python.org/3/library/logging.config.html#logging-config-dictschema>
+# for more complex/consistent logging requirements.
+DICT_CONFIG = {
```
What is this settings file used for? Could it be merged with the previous config file?

awelzel replied:
It's the configuration file of the `rq worker`. I've move this part into config.yml and renamed the file to `rq_worker_settings.py`.

timwoj commented on ci-based/zeek_benchmarker/tasks.py:
```diff
@@ -115,30 +121,29 @@ def unpack_build(
+            working_dir=source_dir,
+            security_opt=["no-new-privileges"],
+            command=["bash", "-x", "-c", command],
+        )
```
I can see this being better than the docker-compose, plus not having to run `subprocess.run` is good.

awelzel replied:
I've re-worked running the benchmarker container using the Docker API directly. A bit of a draw-back is that the image which runs the benchmarks now has to be built, but not the end of the world IMO.

timwoj reviewed (approved):
This looks good to me overall. I had one minor question about configuration. If you want to rework that docker-compose file into the code that's fine with me.

awelzel committed:
docker/docker-compose: Run api, rq and redis in docker-compose

awelzel committed:
config: Include rq dictConfig

awelzel committed:
Run benchmarker containers without docker-compose

We continue to use docker-compose, but it is now running
the infrastructure. Containers are launched from the rq
container via the docker Python library.

Broker benchmarking is now definitely not working.

awelzel committed:
pre-commit: Add yamllint

awelzel committed:
pre-commit: Add hadolint for ci-builder

awelzel committed:
api: Remove reliance on git subprocess

No need to install the git package for checking the validity of
a branch name.

awelzel committed:
pyupgrade: Use --py310-plus, nicer type annotations

Commit:
Fixup after first live deployment

awelzel committed:
Update README.md a bit

awelzel committed:
docker-compose: Gunicorn access logs to stdout

awelzel commented:
> This looks good to me overall. I had one minor question about configuration. If you want to rework that docker-compose file into the code that's fine with me.

@timwoj - feel free to take another looks. I mentioned the locations of the new database in DM if you want to look at some of the results.
